### ISSUE TITLE
Package for nix

### ISSUE THREAD

**Issue by IceDBorn** [TheAlexDev23/power-options#10]

**Comment by TheAlexDev23** [comment]
I've never used the nix package manager nor NixOS, so I've got 0 experience packaging for nix. If anyone does I'll be happy to accept any PRs but otherwise I don't think this will be a priority.

**Comment by mark-boute** [comment]
I'll give it a try, @TheAlexDev23 could you provide the Cargo.lock with releases?

**Comment by TheAlexDev23** [comment]
@mark-boute yeah sorry, I don't know why I had it excluded from vcs

**Comment by mark-boute** [comment]
Alright, I got stuck at setting up the systemctl service as it is not available. Leaving my attempt here for someone else to use:

```nix
{ lib,    
  libadwaita,
    # yad
  dbus,
  pkg-config,
  atk,
  xdotool,
  gtk3, 
  gtk4, 
  webkitgtk_4_1,
  fetchFromGitHub,
  rustPlatform,
  zsh
}:

rustPlatform.buildRustPackage rec {
  pname = "power-options-gtk";
  version = "591afd6f49ce79cfb5d2d2d1d43e10e64bf44a86";
  # use commit that added Cargo.lock instead of release

  src = fetchFromGitHub {
    owner = "TheAlexDev23";
    repo = "power-options";
    rev = version;
    hash = "sha256-gXSJej3Bl2ZnMl3vlSwaknR1yhlaEsRlGctR8abd+GY=";
  };

  cargoHash = "sha256-QCFtj+zcNJ5AvGq3tpi0BqUSjvDo8o/HA4Bs/3VrTZY=";

  buildInputs = [ 
    libadwaita
    # yad
    dbus
    xdotool
    atk 
    gtk3 
    gtk4 
    webkitgtk_4_1
  ];

  nativeBuildInputs = [ 
    zsh
    pkg-config
   ];

  cargoInstallHook = ''
    mkdir -p $out/bin
    cp target/release/power-daemon-mgr $out/bin/
    cp target/release/frontend-gtk $out/bin/power-options-gtk
    cp icon.png $out/share/icons/power-options.png
    cp install/power-options-gtk.desktop $out/share/applications/
  '';

  meta = {
    description = "A gtk frontend for Power Options, a blazingly fast power management solution.";
    homepage = "https://github.com/TheAlexDev23/power-options";
    license = lib.licenses.mit;
  };
}```

Only the `power-daemon-mgr` seems to be available in `/run/current-system/sw/bin/power-daemon-mgr`

**Comment by TheAlexDev23** [comment]
The installation process requires the call to `power-daemon-mgr generate-base-files` to generate the files for systemd, acpi, dbus, etc. They are not available as files in the repo, mostly because they require some form of string interpolation to be properly generated.

Take a look at the AUR PKGBUILD, https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=power-options-daemon.

Tell me if this is not suitable for nix, and if I should think on alternative installation methods.

**Comment by mark-boute** [comment]
I figured and tried to run this imperatively. Where does it write to?
I think the main limiting factor is me not knowing how I turn this derivation into a systemd service.

**Comment by TheAlexDev23** [comment]
The `generate-base-files` command takes 2 parameters:

- `--program-path` which indicates the final path where the power-daemon-mgr executable will be located
- `--path` which will be treated as the root folder during installation.

It will generate the following files relative to `--path`
- /usr/lib/udev/rules.d/85-power-daemon.rules
- /etc/acpi/events/power-options
- /usr/share/dbus-1/system.d/power-daemon.conf
- /usr/lib/systemd/system/power-options.service

Is this what you were asking?

**Comment by mark-boute** [comment]
Right, thanks. I am currently looking into this again. giving one root for those files may not work in Nix, see: https://discourse.nixos.org/t/after-the-derivation-is-built-what-happens-to-files-in-out/24711
I am still pretty new to Nix so it's a learning experience for me too :)

**Comment by Arthur-Landel** [comment]
I managed to build it with the following (they need to be added with git, committing is not needed).
But I didn't find out, how to add the runtime dependencies.

About the systemd services, maybe this will help you:
https://search.nixos.org/options?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=systemd.services

default.nix
```
{ pkgs ? import <nixpkgs> { } }:
let manifest = (pkgs.lib.importTOML ./crates/power-daemon/Cargo.toml).package;
in
pkgs.rustPlatform.buildRustPackage {
  pname = manifest.name;
  version = manifest.version;
  cargoLock.lockFile = ./Cargo.lock;
  src = pkgs.lib.cleanSource ./.;
  
  nativeBuildInputs = with pkgs; [
    pkg-config
  ];
  buildInputs = with pkgs; [
    libadwaita
    graphene
    glib
    gtk3
    gtk4
    webkitgtk_4_1
    xdotool
  ];
}
```

flake.nix
```
{
  description = "All-In-One blazingly fast Linux GUI Application for simple and advanced power
  management on any device";
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
  };
  outputs = { nixpkgs, ... }:
    let
      supportedSystems = [ "x86_64-linux" ];
      forAllSystems = nixpkgs.lib.genAttrs supportedSystems;
      pkgsFor = nixpkgs.legacyPackages;
    in {
      packages = forAllSystems (system: {
        default = pkgsFor.${system}.callPackage ./. {};
      });
    };
}
```

**Comment by TheAlexDev23** [comment]
Hi @teto, thanks for the interest in power-options. As I have previously said in this chain I have no idea about Nix and its workings, so you might need to help me with some stuff if it is not the Nix way.

> the GUI worked but it crashes with unhelpful errors when it can't read a file : "Couldnot open file" without sharing what file it was tryng to open (same when it can't run "yad")

I tried to trace the GUI error in the source code and there is no single failure point in the GTK frontend that should tell you `Could not open file`, only in the daemon in which case it should tell you the file path and the opening error reason.

`yad` is a required dependency because it is used for modals in the frontend, such as crash modals and error messages.

> The daemon has "setup" and generate-base-files which seem redundant to me.

`generate-base-files` generates files for the daemon to run properly, such as `udev` and `acpi` event listening, plus `dbus` and `systemd` configurations that manage daemon background running and communication channels with the frontend.

`setup` on the other hand is used for generating configuration files and profiles for the daemon, it analyzes the system and generates default profiles based on the available features. 

The idea is that `generate-base-files` is run once on install and `setup` may be ran multiple times to regenerate profiles.

>  I was not sure why --program-path would be different than --path ?

`--prorgam-path` is the full path of the program as some systems install it in different locations, usually it's going to be `/usr/bin/...` while `--path` is the root directory of the system, it will be used as the relative path for the required files to install e.g. `[path]/usr/lib/systemd/system/power-options.service`. The `--path` parameter exists because when setting up the program in install environments, or at least in arch linux you don't install to root directly, you install in a simulated root directory in the package setup path, then the package manager can track all the files related to the package and delete them when needed. `--program-path` on the other hand is the full path of the program after it has been installed and will be used in the generated files to reference the power-options binary.

> Here I dont know where it's trying to write and what. Even with -v, there is no explicit path.

Line 43 corresponds to an attempt to create `[path]/etc/power-options/`.

> here I am not sure why it stops or how to workaround it.

D-Bus is refusing name ownership, given that you had issues running `generate-base-files` the d-bus configuration file might not have been generated. Or maybe the configuration hasn't been applied because you haven't reloaded/restarted d-bus after generating the config file. 

>  I am full wayland so the xset error is normal.

Most likely, although from my limited experience Wayland still has X supported functions and xset errors are quite typical given that the default display environment variable is not standardized across systems.
 
> My network driver seems to be "ath11k_pci" which looks unsupported.

Yes only `iwlmvm` and `iwldvm` are supported.

> I wonder if some ERRORs should not be warnings since the software keeps going ?

Yeah, that is the case. Warnings are usually used for situations where a fallback solution exists, such as, for example, the support for an alternative firmware that system has. Errors on the other hand are unrecoverable and in most cases, those settings will be ignored as the system is an unexpected or unsupported state.
